velox/connectors/hive/storage_adapters/gcs/RegisterGcsFileSystem.cpp

```diff
@@ -32,24 +42,45 @@ gcsFileSystemGenerator() {
   static auto filesystemGenerator =
       [](std::shared_ptr<const config::ConfigBase> properties,
          std::string_view filePath) {
-        // Only one instance of GCSFileSystem is supported for now (follow S3
-        // for now).
-        // TODO: Support multiple GCSFileSystem instances using a cache
-        // Initialize on first access and reuse after that.
-        static std::shared_ptr<FileSystem> gcsfs;
-        folly::call_once(GcsInstantiationFlag, [&properties]() {
-          std::shared_ptr<GcsFileSystem> fs;
-          if (properties != nullptr) {
-            fs = std::make_shared<GcsFileSystem>(properties);
-          } else {
-            fs = std::make_shared<GcsFileSystem>(
-                std::make_shared<config::ConfigBase>(
-                    std::unordered_map<std::string, std::string>()));
-          }
-          fs->initializeClient();
-          gcsfs = fs;
-        });
-        return gcsfs;
+        const auto file = gcsPath(filePath);
+        std::string bucket;
+        std::string object;
+        setBucketAndKeyFromGcsPath(file, bucket, object);
+
+        // Check if an instance exists with a read lock (shared).
+        auto fs = fileSystems().withRLock(
+            [&](auto& instanceMap) -> std::shared_ptr<FileSystem> {
+              auto iterator = instanceMap.find(bucket);
```

Comment:
Seems we should use endpoint as key if found, which means each endpoint has only one filesystem created.

Reply:
@PHILO-HE Thanks for your pointing. I have fixed. Can you help to review again? Thanks.